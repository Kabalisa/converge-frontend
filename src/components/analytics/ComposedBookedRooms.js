import React from 'react';
import { PropTypes } from 'prop-types';
import '../../assets/styles/composed-rooms.scss';
import QueryMostBookedRooms from './QueryMostBookedRooms';
import QueryLeastBookedRooms from './QueryLeastBookedRooms';

/**
 * Component for composed booked rooms
 *
 * @param {object} dateValueObject
 *
 * @returns {JSX}
 */
const ComposedBookedRooms = ({ dateValue }) => (
  <div className="wrap-composed-rooms">
    <div>
      {
        <QueryMostBookedRooms dateValue={dateValue} />
      }
    </div>
    <div id="booked-room-margin">
      {
        <QueryLeastBookedRooms dateValue={dateValue} />
      }
    </div>
  </div>
);
ComposedBookedRooms.propTypes = {
  dateValue: PropTypes.shape({
    startDate: PropTypes.string,
    endDate: PropTypes.string,
  }),
};

ComposedBookedRooms.defaultProps = {
  dateValue: {},
};
export default ComposedBookedRooms;
