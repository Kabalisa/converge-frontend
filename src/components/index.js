import NavBar from './NavBar';
import SideBar from './SideBar';
import ContentBlock from './ContentBlock';
import Login from './Login';

export { NavBar, SideBar, ContentBlock, Login };
