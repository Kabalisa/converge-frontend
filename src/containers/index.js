import Settings from './Settings';
import Analytics from './Analytics';
import Preference from './Preference';
import RoomFeedbackPage from './RoomFeedbackPage';

export { Settings, Analytics, Preference, RoomFeedbackPage };
