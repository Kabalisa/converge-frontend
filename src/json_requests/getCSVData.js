import axios from 'axios';
import { getToken } from '../utils/Cookie';

const token = getToken();
const { BASE_URL } = process.env;
export const apiRequest = axios.create({
  baseURL: BASE_URL,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    Authorization: `Bearer ${token}`,
  },
});

  /**
   * queries csv data
   * @param {string} startDate - start date
   * @param {string} endDate - end date
   */
  /* istanbul ignore next */
const getCSVData = (startDate, endDate) =>
  apiRequest.post('', {
    start_date: startDate,
    end_date: endDate,
    file_type: 'csv',
  });
export default getCSVData;
