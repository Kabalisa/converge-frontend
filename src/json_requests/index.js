import axios from 'axios';
import { getItemFromLocalStorage } from '../utils/Utilities';

const token = getItemFromLocalStorage('MRM_AUTH_TOKEN');
const { BASE_URL } = process.env;
export const apiRequest = axios.create({
  baseURL: BASE_URL,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    Authorization: `Bearer ${token}`,
  },
});
export const getMostUsedAndLeastUsedRooms = (startDate, endDate) =>
  apiRequest.post('', {
    start_date: startDate,
    end_date: endDate,
  });
