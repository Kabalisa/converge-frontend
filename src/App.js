import React, { Component } from 'react';
import { withRouter, Route, Switch, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import ROUTES from './utils/routes';
import { Settings, Analytics, Feedback } from './containers';
import { LoginPage } from './components';
import { getItemFromLocalStorage } from './utils/Utilities';
import Constants from './utils/Constants';
import '../src/assets/styles/toastr.scss';

// destruscture constants to be used
const {
  MRM_TOKEN,
  MESSAGES: { authenticationError },
} = Constants;
class App extends Component {
  static propTypes = {
    location: PropTypes.shape({
      pathname: PropTypes.string,
    }).isRequired,
  };

  static getDerivedStateFromProps = () => {
    const token = getItemFromLocalStorage(MRM_TOKEN);
    if (!token) {
      return { loggedIn: false };
    }
    return { loggedIn: true };
  };

  state = {
    loggedIn: false,
  };

  render() {
    const { loggedIn } = this.state;
    const { location } = this.props;

    if (!loggedIn && location.pathname !== ROUTES.home) {
      // redirect to landing page if user isn't logged in
      return (
        <Redirect
          to={{
            pathname: ROUTES.home,
            state: { errorMessage: authenticationError },
          }}
        />
      );
    }

    return (
      <Switch>
        <Route exact path={ROUTES.home} component={LoginPage} />
        <Route path={ROUTES.settings} component={Settings} />
        <Route path={ROUTES.analytics} component={Analytics} />
        <Route path={ROUTES.feedback} component={Feedback} />
      </Switch>
    );
  }
}

export default withRouter(App);
