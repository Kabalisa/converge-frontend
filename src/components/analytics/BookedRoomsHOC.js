import React from 'react';
import { PropTypes } from 'prop-types';
import PollIcon from '../../assets/images/poll.svg';
import RedPollIcon from '../../assets/images/poll_red.svg';
import MoreIcon from '../../assets/images/more.svg';
// import leastBookedRooms from '../../fixtures/leastBookedRooms';
import mostBookedRooms from '../../fixtures/mostBookedRooms';
import { getMostUsedAndLeastUsedRooms } from '../../json_requests';
import { thisWeek } from '../../utils/Utilities';

export class ComposedBooked extends React.Component {
  state = {
    leastUsedRooms: [],
    fetching: false,
    // eslint-disable-next-line
    error: null,
  };
  componentWillReceiveProps({ dateValue }) {
    this.fetchMostAndLeastUsedRooms(dateValue);
  }
  fetchMostAndLeastUsedRooms = (dateValue) => {
    this.setState({ fetching: true });
    let startDate;
    let endDate;
    if (dateValue === 'This Week') {
      const dates = thisWeek();
      startDate = dates.weekStart;
      endDate = dates.weekEnd;
    }

    return (
      getMostUsedAndLeastUsedRooms(startDate, endDate)
        .then((response) => {
          const rooms = response.data['Least Used Rooms'].Room;
          const meetings = response.data['Least Used Rooms'].Meetings;
          const meetingShares =
            response.data['Least Used Rooms']['% Share of All Meetings'];
          this.setState({
            leastUsedRooms: [rooms, meetings, meetingShares],
            fetching: false,
          });
        })
        // this fetching value is not set to false to avoid props validation
        // failure in the child component, otherwise it should be set to false
        // eslint-disable-next-line
        .catch((error) => this.setState({ fetching: true, error }))
    );
  };
  render() {
    const ComposedBookedRooms = this.props.component;
    if (this.props.bookedRoomText === 'Most Booked Rooms') {
      return (
        <ComposedBookedRooms
          pollIcon={PollIcon}
          moreIcon={MoreIcon}
          bookedRoomText={this.props.bookedRoomText}
          fetching={false}
          bookedRoomsList={mostBookedRooms}
        />
      );
    }
    return (
      <ComposedBookedRooms
        pollIcon={RedPollIcon}
        moreIcon={MoreIcon}
        bookedRoomText={this.props.bookedRoomText}
        fetching={this.state.fetching}
        bookedRoomsList={this.state.leastUsedRooms}
      />
    );
  }
}

ComposedBooked.propTypes = {
  dateValue: PropTypes.string.isRequired,
  component: PropTypes.func.isRequired,
  bookedRoomText: PropTypes.string.isRequired,
};

export default ComposedBooked;
