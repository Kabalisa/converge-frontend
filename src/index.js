import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import { ThemeProvider } from 'react-css-themr';
import { ApolloProvider } from 'react-apollo';
import App from './App';
import apolloClient from './utils/ApolloClient';
import './assets/styles/index.scss';

const RTButton = require('react-toolbox/lib/button/theme.css');
const RTDialog = require('react-toolbox/lib/dialog/theme.css');
const RTLink = require('react-toolbox/lib/link/theme.css');
const RTSnackbar = require('react-toolbox/lib/snackbar/theme.css');
const RTDropdown = require('react-toolbox/lib/dropdown/theme.css');

const contextTheme = {
  RTButton,
  RTDialog,
  RTLink,
  RTSnackbar,
  RTDropdown,
};

ReactDOM.render(
  <ThemeProvider theme={contextTheme}>
    <ApolloProvider client={apolloClient}>
      <Router>
        <App />
      </Router>
    </ApolloProvider>
  </ThemeProvider>,
  document.getElementById('root'),
);
