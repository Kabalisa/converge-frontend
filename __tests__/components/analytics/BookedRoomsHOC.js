import React from 'react';
import { shallow } from 'enzyme';
import MockAdapter from 'axios-mock-adapter';
import BookedRoomsHOC from '../../../src/components/analytics/BookedRoomsHOC';
import { apiRequest } from '../../../src/json_requests';
import ComposedBookedRooms from '../../../src/components/analytics/ComposedBookedRooms';
import leastUsedRooms from '../../../__mocks__/rooms/leastUsedRooms';

const mock = new MockAdapter(apiRequest);
const { BASE_URL } = process.env || '';

const props = {
  dateValue: 'Today',
  component: ComposedBookedRooms,
  bookedRoomText: '',
};

describe('BookedRoomsHOC Component', () => {
  it('should renders correctly', () => {
    const MostBookedRooms = BookedRoomsHOC;
    const LeastBookedRooms = BookedRoomsHOC;

    expect(shallow(<MostBookedRooms {...props} />)).toMatchSnapshot();
    expect(shallow(<LeastBookedRooms {...props} />)).toMatchSnapshot();
    expect(shallow(<ComposedBookedRooms />)).toMatchSnapshot();
  });

  it('should have empty list initial state', () => {
    const component = shallow(<BookedRoomsHOC {...props} />);
    expect(component.state()).toEqual({
      leastUsedRooms: [],
      fetching: false,
      error: null,
    });
  });
});

describe('Updates the state after fetching the API data', () => {
  const response = leastUsedRooms;
  mock.onPost(BASE_URL).reply(200, response.data);

  it('should update the state with least used rooms (This Week)', async () => {
    const wrapperWeekly = shallow(<BookedRoomsHOC {...props} />);
    props.dateValue = 'This Week';
    const { dateValue } = props;

    await wrapperWeekly.instance().fetchMostAndLeastUsedRooms(dateValue);
    const rooms = response.data['Least Used Rooms'].Room;
    const meetings = response.data['Least Used Rooms'].Meetings;
    const meetingShares =
      response.data['Least Used Rooms']['% Share of All Meetings'];
    const data = [rooms, meetings, meetingShares];

    expect(wrapperWeekly.state()).toEqual({
      leastUsedRooms: data,
      fetching: false,
      error: null,
    });
  });
  it('should update the state with an error when one occurs', async () => {
    mock.onPost(BASE_URL).reply(500);
    const wrapperDaily = shallow(<BookedRoomsHOC {...props} />);
    await wrapperDaily.instance().fetchMostAndLeastUsedRooms();
    expect(wrapperDaily.state('error')).not.toBe(null);
  });
});
