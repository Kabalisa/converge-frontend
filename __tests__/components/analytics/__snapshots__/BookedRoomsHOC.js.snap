// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookedRoomsHOC Component should renders correctly 1`] = `
<ComposedBookedRooms
  bookedRoomText=""
  bookedRoomsList={Array []}
  dateValue="Today"
  fetching={false}
  moreIcon="test-file-stub"
  pollIcon="test-file-stub"
/>
`;

exports[`BookedRoomsHOC Component should renders correctly 2`] = `
<ComposedBookedRooms
  bookedRoomText=""
  bookedRoomsList={Array []}
  dateValue="Today"
  fetching={false}
  moreIcon="test-file-stub"
  pollIcon="test-file-stub"
/>
`;

exports[`BookedRoomsHOC Component should renders correctly 3`] = `
<div
  className="wrap-composed-rooms"
>
  <div>
    <ComposedBooked
      bookedRoomText="Most Booked Rooms"
      component={[Function]}
      dateValue="Today"
    />
  </div>
  <div
    id="booked-room-margin"
  >
    <ComposedBooked
      bookedRoomText="Least Booked Rooms"
      component={[Function]}
      dateValue="Today"
    />
  </div>
</div>
`;
